README.md
# Logstash Plugin

[![Travis Build Status](https://travis-ci.org/logstash-plugins/logstash-output-statsd.svg)](https://travis-ci.org/logstash-plugins/logstash-output-statsd)

This is a plugin for [Logstash](https://github.com/elastic/logstash).

It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.

## Documentation

Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).

- For formatting code or config example, you can use the asciidoc `[source,ruby]` directive
- For more asciidoc formatting tips, see the excellent reference here https://github.com/elastic/docs#asciidoc-guide

## Need Help?

Need help? Try #logstash on freenode IRC or the https://discuss.elastic.co/c/logstash discussion forum.

## Developing

### 1. Plugin Developement and Testing

#### Code
- To get started, you'll need JRuby with the Bundler gem installed.

- Create a new plugin or clone and existing from the GitHub [logstash-plugins](https://github.com/logstash-plugins) organization. We also provide [example plugins](https://github.com/logstash-plugins?query=example).

- Install dependencies
```sh
bundle install
```

#### Test

- Update your dependencies

```sh
bundle install
```

- Run tests

```sh
bundle exec rspec
```

### 2. Running your unpublished Plugin in Logstash

#### 2.1 Run in a local Logstash clone

- Edit Logstash `Gemfile` and add the local plugin path, for example:
```ruby
gem "logstash-filter-awesome", :path => "/your/local/logstash-filter-awesome"
```
- Install plugin
```sh
# Logstash 2.3 and higher
bin/logstah-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Run Logstash with your plugin
```sh
bin/logstash -e 'filter {awesome {}}'
```
At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.

#### 2.2 Run in an installed Logstash

You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:

- Build your plugin gem
```sh
gem build logstash-filter-awesome.gemspec
```
- Install the plugin from the Logstash home
```sh
# Logstash 2.3 and higher
bin/logstah-plugin install --no-verify

# Prior to Logstash 2.3
bin/plugin install --no-verify

```
- Start Logstash and proceed to test the plugin

## Contributing

All contributions are welcome: ideas, patches, documentation, bug reports, complaints, and even something you drew up on a napkin.

Programming is not a required skill. Whatever you've seen about open source and maintainers or community members  saying "send patches or die" - you will not see that here.

It is more important to the community that you are able to contribute.

For more information about contributing, see the [CONTRIBUTING](https://github.com/elastic/logstash/blob/master/CONTRIBUTING.md) file.

CHANGELOG.md
## 3.2.0
  - Added TCP support [30](https://github.com/logstash-plugins/logstash-output-statsd/pull/30)
  - Changed test code for cleanups [31](https://github.com/logstash-plugins/logstash-output-statsd/pull/31)
  
## 3.1.5
  - Docs: Set the default_codec doc attribute.

## 3.1.4
  - Update gemspec summary

## 3.1.3
  - Fix some documentation issues

## 3.1.1
  - Relax constraint on logstash-core-plugin-api to >= 1.60 <= 2.99

## 3.1.0
 - breaking,config: Remove deprecated config `debug`.

## 3.0.1
 - Republish all the gems under jruby.

## 3.0.0
 - Update the plugin to the version 2.0 of the plugin api, this change is required for Logstash 5.0 compatibility. See https://github.com/elastic/logstash/issues/5141

## 2.0.7
 - Depend on logstash-core-plugin-api instead of logstash-core, removing the need to mass update plugins on major releases of logstash

## 2.0.6
 - New dependency requirements for logstash-core for the 5.0 release

## 2.0.5
 - Doc improvements.

## 2.0.2
 - Avoid rspec.configure global after and before :all blocks

## 2.0.0
 - Plugins were updated to follow the new shutdown semantic, this mainly allows Logstash to instruct input plugins to terminate gracefully,
   instead of using Thread.raise on the plugins' threads. Ref: https://github.com/elastic/logstash/pull/3895
 - Dependency on logstash-core update to 2.0

## 1.1.0
 - improved test design to be more rspec3 friendly, including the usage
    of random ports to avoid colisions in integration test.


spec/spec_helper.rb
require "logstash/devutils/rspec/spec_helper"
require "socket"

class StatsdServer

  attr_reader :received, :port

  def initialize
    @sync_lock = Mutex.new
    @terminated = false
    @received   = []
  end

  def register(port, protocol)
    @port = port
    if protocol == "udp"
      @socket = UDPSocket.new
      @socket.bind("127.0.0.1", port)
    else
      @socket = TCPserver.new("127.0.0.1", port)
    end
  end


  def run(port, protocol = "udp")
    register(port, protocol)

    Thread.new do
      if protocol == "udp"
        udp_receive
      else
        tcp_receive
      end
    end

    self
  end

  def append(metric)
    @sync_lock.synchronize do
      @received << metric
    end
  end

  def close
    @sync_lock.synchronize do
      @terminated = true
    end
  end

  def closed?
    @terminated == true
  end
end

private

def tcp_receive
  client = @socket.accept
  Timeout.timeout(5) { sleep(0.1) while client.nil? }
  metric = client.recvfrom(100).first
  append(metric)
  client.close
end

def udp_receive
  while(!closed?)
    metric, _ = @socket.recvfrom(100)
    append(metric)
  end
end

RSpec.configure do |c|
  srand(c.seed)
end


spec/outputs/statsd_spec.rb
# encoding: utf-8
require_relative "../spec_helper"
require "logstash/outputs/statsd"
require "socket"

describe LogStash::Outputs::Statsd do

  describe "registration and close" do
    it "should register without errors" do
      output = LogStash::Plugin.lookup("output", "statsd").new
      expect {output.register}.to_not raise_error
    end
  end

  describe "IO" do
    let(:host) { "localhost" }
    let(:port) { rand(2000..10000) }
    let!(:server) { StatsdServer.new.run(port, protocol) }

    after(:each) { server.close }

    shared_examples "it receives sent data" do
      let(:config) do
        { "host" => host, "sender" => "spec", "port" => port, "protocol" => protocol, "count" => [ "foo.bar", "0.1" ] }
      end
      let(:properties) do
        { "metric" => "foo.bar", "count" => 10 }
      end
      let(:event) { LogStash::Event.new(properties) }

      subject { LogStash::Outputs::Statsd.new(config) }

      before(:each) { subject.register }

      it "should receive data send to the server" do
        subject.receive(event)
        # Since we are dealing with threads and networks,
        # we might experience delays or timing issues.
        # lets try a few times before giving up completely.
        try { expect(server.received).to include("logstash.spec.foo.bar:0.1|c") }
      end
    end

    describe "UDP" do
      let(:protocol) { "udp" }

      context "#send" do
        include_examples "it receives sent data"
      end
    end

    describe "TCP" do
      let(:protocol) { "tcp" }

      context "#send" do
        include_examples "it receives sent data"
      end
    end
  end
end


spec/outputs/dogstatsd_spec.rb
# encoding: utf-8
require_relative '../spec_helper'

describe LogStash::Outputs::Dogstatsd do
  let(:output) { described_class.new(config) }

  let(:config) do
    {
      'host' => '127.0.0.1',
      'port' => 8125
    }.merge(metric_config)
  end
  let(:metric_config) { {} }

  describe 'registration and close' do
    it 'registers without errors' do
      output = LogStash::Plugin.lookup('output', 'dogstatsd').new
      expect { output.register }.to_not raise_error
    end
  end

  describe '#send' do
    before { output.register }
    subject { output.receive(LogStash::Event.new(event)) }

    let(:event) { { 'something_count' => 10 } }

    context 'increment metrics' do
      let(:metric_config) { { 'increment' => [metric_to_track] } }
      let(:metric_to_track) { 'metric.name.here' }

      context 'with a plain ol metric name' do
        it 'tracks' do
          expect_any_instance_of(Datadog::Statsd).to receive(:send_to_socket)
            .with("#{metric_to_track}:1|c")
          subject
        end
      end

      context 'with tags' do
        let(:metric_config) { super().merge('metric_tags' => ['foo:%{value}']) }
        let(:event) { { 'value' => 'helloworld' } }

        it 'sprintf tags' do
          expect_any_instance_of(Datadog::Statsd).to receive(:send_to_socket)
            .with("#{metric_to_track}:1|c|#foo:helloworld")
          subject
        end
      end
    end

    context 'histogram metrics' do
      let(:metric_to_track) { 'metric.name.here' }
      let(:metric_config) { { 'histogram' => { '%{metric_name}' => '%{track_value}' } } }
      let(:event) { super().merge('metric_name' => metric_to_track, 'track_value' => 123) }

      context 'with event fields in the metric name and value' do
        it 'tracks' do
          expect_any_instance_of(Datadog::Statsd).to receive(:send_to_socket)
            .with("#{metric_to_track}:123|h")
          subject
        end
      end
    end
  end
end


lib/logstash/outputs/dogstatsd.rb
# encoding: utf-8
require "logstash/outputs/base"
require "logstash/namespace"
require "datadog/statsd"

# dogstatsd is a fork of the statsd protocol which aggregates statistics, such
# as counters and timers, and ships them over UDP to the dogstatsd-server
# running as part of the Datadog Agent. Dogstatsd adds support for metric tags,
# which are used to slice metrics along various dimensions.
#
# You can learn about statsd here:
#
# * https://codeascraft.com/2011/02/15/measure-anything-measure-everything/[Etsy blog post announcing statsd]
# * https://github.com/etsy/statsd[statsd on github]
#
# Typical examples of how this can be used with Logstash include counting HTTP hits
# by response code, summing the total number of bytes of traffic served, and tracking
# the 50th and 95th percentile of the processing time of requests.
#
# Example:
# [source,ruby]
# output {
#   dogstatsd {
#     metric_tags => ["host:%{host}","role:foo"]
#     count => {
#       "http.bytes" => "%{bytes}"
#     }
#   }
# }
class LogStash::Outputs::Dogstatsd < LogStash::Outputs::Base
  ## Regex stolen from statsd code
  RESERVED_CHARACTERS_REGEX = /[\:\|\@]/
  config_name "dogstatsd"

  # The hostname or IP address of the dogstatsd server.
  config :host, :validate => :string, :default => "localhost"

  # The port to connect to on your dogstatsd server.
  config :port, :validate => :number, :default => 8125

  # An increment metric. Metric names as array. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :increment, :validate => :array, :default => []

  # A decrement metric. Metric names as array. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :decrement, :validate => :array, :default => []

  # A histogram metric, which a statsd timing but conceptually maps to any
  # numeric value, not just durations. `metric_name => value` as hash. `%{fieldname}`
  # substitutions are allowed in the metric names.
  config :histogram, :validate => :hash, :default => {}

  # A count metric. `metric_name => count` as hash. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :count, :validate => :hash, :default => {}

  # A set metric. `metric_name => "string"` to append as hash. `%{fieldname}`
  # substitutions are allowed in the metric names.
  config :set, :validate => :hash, :default => {}

  # A gauge metric. `metric_name => gauge` as hash. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :gauge, :validate => :hash, :default => {}

  # The sample rate for the metric.
  config :sample_rate, :validate => :number, :default => 1

  # The tags to apply to each metric.
  config :metric_tags, :validate => :array, :default => []

  public
  def register
    @client = Datadog::Statsd.new(@host, @port)
  end # def register

  public
  def receive(event)
    @logger.debug? and @logger.debug("Event: #{event}")

    metric_opts = {
      :sample_rate => @sample_rate,
      :tags => @metric_tags.map { |t| event.sprintf(t) }
    }

    @increment.each do |metric|
      @client.increment(event.sprintf(metric), metric_opts)
    end

    @decrement.each do |metric|
      @client.decrement(event.sprintf(metric), metric_opts)
    end

    @count.each do |metric, val|
      @client.count(event.sprintf(metric), event.sprintf(val), metric_opts)
    end

    @histogram.each do |metric, val|
      @client.histogram(event.sprintf(metric), event.sprintf(val), metric_opts)
    end

    @set.each do |metric, val|
      @client.set(event.sprintf(metric), event.sprintf(val), metric_opts)
    end

    @gauge.each do |metric, val|
      @client.gauge(event.sprintf(metric), event.sprintf(val), metric_opts)
    end
  end # def receive

  public
  def close
    @client.close
  end # def close
end # class LogStash::Outputs::Statsd


lib/logstash/outputs/statsd.rb
# encoding: utf-8
require "logstash/outputs/base"
require "logstash/namespace"

# statsd is a network daemon for aggregating statistics, such as counters and timers,
# and shipping over UDP to backend services, such as Graphite or Datadog. The general
# idea is that you send metrics to statsd and every few seconds it will emit the
# aggregated values to the backend. Example aggregates are sums, average and maximum
# values, their standard deviation, etc. This plugin makes it easy to send such
# metrics based on data in Logstash events.
#
# You can learn about statsd here:
#
# * https://codeascraft.com/2011/02/15/measure-anything-measure-everything/[Etsy blog post announcing statsd]
# * https://github.com/etsy/statsd[statsd on github]
#
# Typical examples of how this can be used with Logstash include counting HTTP hits
# by response code, summing the total number of bytes of traffic served, and tracking
# the 50th and 95th percentile of the processing time of requests.
#
# Each metric emitted to statsd has a dot-separated path, a type, and a value. The
# metric path is built from the `namespace` and `sender` options together with the
# metric name that's picked up depending on the type of metric. All in all, the
# metric path will follow this pattern:
#
#     namespace.sender.metric
#
# With regards to this plugin, the default namespace is "logstash", the default
# sender is the `host` field, and the metric name depends on what is set as the
# metric name in the `increment`, `decrement`, `timing`, `count`, `set` or `gauge`
# options. In metric paths, colons (":"), pipes ("|") and at signs ("@") are reserved
# and will be replaced by underscores ("_").
#
# Example:
# [source,ruby]
# output {
#   statsd {
#     host => "statsd.example.org"
#     count => {
#       "http.bytes" => "%{bytes}"
#     }
#   }
# }
#
# If run on a host named hal9000 the configuration above will send the following
# metric to statsd if the current event has 123 in its `bytes` field:
#
#     logstash.hal9000.http.bytes:123|c
class LogStash::Outputs::Statsd < LogStash::Outputs::Base
  ## Regex stolen from statsd code
  RESERVED_CHARACTERS_REGEX = /[\:\|\@]/
  config_name "statsd"

  # The hostname or IP address of the statsd server.
  config :host, :validate => :string, :default => "localhost"

  # The port to connect to on your statsd server.
  config :port, :validate => :number, :default => 8125

  # The protocol to use for the connection.
  config :protocol, :validate => ['udp', 'tcp'], :default => "udp"

  # The statsd namespace to use for this metric. `%{fieldname}` substitutions are
  # allowed.
  config :namespace, :validate => :string, :default => "logstash"

  # The name of the sender. Dots will be replaced with underscores. `%{fieldname}`
  # substitutions are allowed.
  config :sender, :validate => :string, :default => "%{host}"

  # An increment metric. Metric names as array. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :increment, :validate => :array, :default => []

  # A decrement metric. Metric names as array. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :decrement, :validate => :array, :default => []

  # A timing metric. `metric_name => duration` as hash. `%{fieldname}` substitutions
  # are allowed in the metric names.
  config :timing, :validate => :hash, :default => {}

  # A count metric. `metric_name => count` as hash. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :count, :validate => :hash, :default => {}

  # A set metric. `metric_name => "string"` to append as hash. `%{fieldname}`
  # substitutions are allowed in the metric names.
  config :set, :validate => :hash, :default => {}

  # A gauge metric. `metric_name => gauge` as hash. `%{fieldname}` substitutions are
  # allowed in the metric names.
  config :gauge, :validate => :hash, :default => {}

  # The sample rate for the metric.
  config :sample_rate, :validate => :number, :default => 1

  public
  def register
    require "statsd"
    @client = Statsd.new(@host, @port, @protocol.to_sym)
  end # def register

  public
  def receive(event)
    
    @client.namespace = event.sprintf(@namespace) if not @namespace.empty?
    @logger.debug? and @logger.debug("Original sender: #{@sender}")
    sender = event.sprintf(@sender)
    @logger.debug? and @logger.debug("Munged sender: #{sender}")
    @logger.debug? and @logger.debug("Event: #{event}")
    @increment.each do |metric|
      @client.increment(build_stat(event.sprintf(metric), sender), @sample_rate)
    end
    @decrement.each do |metric|
      @client.decrement(build_stat(event.sprintf(metric), sender), @sample_rate)
    end
    @count.each do |metric, val|
      @client.count(build_stat(event.sprintf(metric), sender),
                    event.sprintf(val), @sample_rate)
    end
    @timing.each do |metric, val|
      @client.timing(build_stat(event.sprintf(metric), sender),
                     event.sprintf(val), @sample_rate)
    end
    @set.each do |metric, val|
      @client.set(build_stat(event.sprintf(metric), sender),
                    event.sprintf(val), @sample_rate)
    end
    @gauge.each do |metric, val|
      @client.gauge(build_stat(event.sprintf(metric), sender),
                    event.sprintf(val), @sample_rate)
    end
  end # def receive

  def build_stat(metric, sender=@sender)
    sender = sender.to_s.gsub('::','.')
    sender.gsub!(RESERVED_CHARACTERS_REGEX, '_')
    sender.gsub!(".", "_")
    metric = metric.to_s.gsub('::','.')
    metric.gsub!(RESERVED_CHARACTERS_REGEX, '_')
    @logger.debug? and @logger.debug("Formatted value", :sender => sender, :metric => metric)
    return "#{sender}.#{metric}"
  end
end # class LogStash::Outputs::Statsd
